src/agent.vala

```diff
@@ -0,0 +1,186 @@
+        pin_action = new GLib.SimpleAction.stateful ("pin", null, new Variant.string (""));
+        pin_action.change_state.connect ((value) => {
+            this.passkey = value.get_string ();
+        });
+        actions.add_action (pin_action);
+
+        loop = new MainLoop (null, false);
+        bluetooth = bluez;
+        Notify.init ("ayatana-indicator-bluetooth");
+    }
+
+    /* TODO: Add a better way to differentiate between rejected and cancelled errors, maybe with an enum */
+    private bool sendNotification (string device_name, string body, bool need_input, bool have_actions)
+    {
+        bool accepted = !have_actions;
+
+        notification = new Notify.Notification (@"Pair with $device_name?", body, "bluetooth-active");
+        notification.closed.connect (() => {
+            accepted = false;
+            notification = null;
+
+            if (loop.is_running ()) {
+                loop.quit ();
+            }
+        });
+
+        bool is_lomiri = AyatanaCommon.utils_is_lomiri ();
+
+        if (is_lomiri) {
+            if (have_actions) {
+                notification.set_hint ("x-lomiri-snap-decisions", true);
+                notification.set_hint ("x-lomiri-private-affirmative-tint", "true");
+            }
+
+            if (need_input) {
+                VariantBuilder actions_builder = new VariantBuilder (new VariantType ("a{sv}"));
+                actions_builder.add ("{sv}", "notifications", new Variant.string (actions_path));
+
+                VariantBuilder builder = new VariantBuilder (new VariantType ("a{sv}"));
+                builder.add ("{sv}", "busName", new Variant.string ("org.ayatana.indicator.bluetooth"));
+                builder.add ("{sv}", "menuPath", new Variant.string (menu_path));
+                builder.add ("{sv}", "actions", actions_builder.end ());
+
+                notification.set_hint ("x-lomiri-private-menu-model", builder.end ());
+            }
+        }
+
+        if (have_actions) {
+            notification.add_action("yes_id", "Yes", (notif, action) => {
+                loop.quit ();
+                notification = null;
+                accepted = true;
+            });
+            notification.add_action("no_id", "No", (notif, action) => {
+                loop.quit ();
+                notification = null;
+                accepted = false;
+            });
+        }
+
+        if (!have_actions && !need_input) {
+            // Display-only notification. Make sure we don't time out.
+            notification.set_hint ("urgency", 2);
+        }
+
+        try {
+            notification.show ();
+        }
+        catch (Error e) {
+            warning ("Panic: Failed showing notification: %s", e.message);
+        }
+
+        if (have_actions) {
+            loop.run ();
+        }
+
+        return accepted;
+    }
+
+    public void AuthorizeService (GLib.ObjectPath object, string uuid) throws GLib.DBusError, GLib.IOError
+    {
+    }
```

Comment:
As it is, the (background) agent will now authorize any connection request from the other device, negating Bluez' "trusted" flag ("Connect automatically when detected" in lomiri-system-settings). Is that intended?